libsolidity/codegen/Compiler.cpp

```diff
@@ -37,14 +37,16 @@ void Compiler::compileContract(
 	bytes const& _metadata
 )
 {
-	auto static notTransient = [](VariableDeclaration const* _varDeclaration) {
+	auto static notTransientReferenceType = [](VariableDeclaration const* _varDeclaration) {
 		solAssert(_varDeclaration);
-		return _varDeclaration->referenceLocation() != VariableDeclaration::Location::Transient;
+		return
+			_varDeclaration->referenceLocation() != VariableDeclaration::Location::Transient ||
+			!_varDeclaration->hasReferenceOrMappingType();
 	};
 
 	solUnimplementedAssert(
-		ranges::all_of(_contract.stateVariables(), notTransient),
-		"Transient storage variables are not supported."
+		ranges::all_of(_contract.stateVariables(), notTransientReferenceType),
```

Comment:
Changing the lambda to check for the inverse - i.e. whether a variable is transient *and* a reference type, and then using `ranges::any_of` instead would likely be a better option here (and more efficient).

Reply:
While we're at it, it might be better to use `Type::isValueType()` rather than `hasReferenceOrMappingType()` for the check. We do have types that are neither reference nor values types. None of them can be named and it should not be possible to use them in a state variable declaration, but just in case there's some odd corner case we're missing it would be safer for the check to only allow value types and nothing else.

Reply:
I used `none_of` instead of `any_of`.